task/queue.go

```diff
@@ -0,0 +1,156 @@
+package task
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+)
+
+// Queue is handling list of processes and makes sure
+// only one process is executed at the time
+type Queue struct {
+	*sync.Mutex
+	work  chan *Task
+	tasks []*Task
+	// wait group to be able to close queue
+	wgQueue sync.WaitGroup
+	// wait group for tasks to finish
+	wgtasks   sync.WaitGroup
+	idCounter int
+}
+
+// NewQueue creates empty queue ready to be tasked
+func NewQueue() *Queue {
+	q := &Queue{
+		Mutex: &sync.Mutex{},
+		work:  make(chan *Task),
+		tasks: make([]*Task, 0),
+	}
+
+	// Start single worker for queue
+	q.wgQueue.Add(1)
+	go func() {
+		for {
+			task, ok := <-q.work
+			if !ok {
+				q.wgQueue.Done()
+				return
+			}
+
+			q.Lock()
+			task.State = RUNNING
+			q.Unlock()
+
+			err := task.process(task.output)
+
+			q.Lock()
+
+			if err != nil {
+				fmt.Fprintf(task.output, "Task failed with error: %v", err)
+				task.State = FAILED
+			} else {
+				fmt.Fprintln(task.output, "Task succeeded")
+				task.State = SUCCEEDED
+			}
+
+			q.Unlock()
+			q.wgtasks.Done()
+		}
+
+	}()
+
+	return q
+}
+
+// GetTasks gets complete list of tasks
+func (q *Queue) GetTasks() []Task {
+	var tasks []Task
+	q.Lock()
+	for _, task := range q.tasks {
+		tasks = append(tasks, *task)
+	}
+
+	q.Unlock()
+	return tasks
+}
+
+// GetTaskByID returns task with given id
+func (q *Queue) GetTaskByID(ID int) (Task, error) {
+	q.Lock()
+	tasks := q.tasks
+	q.Unlock()
+
+	for _, task := range tasks {
+		if task.ID == ID {
+			return *task, nil
+		}
+	}
```

Comment:
I assume linear lookup is fine here as we're not expecting too many tasks... However maybe it'd be a good idea still to store them in a by-id hashmap/similar?